sympy/ntheory/factor_.py

```diff
@@ -28,6 +28,17 @@
  216091, 756839, 859433, 1257787, 1398269, 2976221, 3021377, 6972593, 13466917, 20996011, 24036583,
  25964951, 30402457, 32582657, 37156667, 42643801, 43112609, 57885161, 74207281, 77232917, 82589933)
 
+PERFECT = MERSENNES = None  # compute lazily
+def _ismersenneprime(n):
+    global MERSENNES
+    if MERSENNES is None:
+        MERSENNES = [2**i - 1 for i in MERSENNE_PRIME_EXPONENTS]
+    return n in MERSENNES
+def _isperfect(n):
+    global PERFECT
+    if PERFECT is None:
+        PERFECT = [2**(i - 1)*(2**i - 1) for i in MERSENNE_PRIME_EXPONENTS]
+    return n in PERFECT
```

Comment:
This takes a large amount of memory- these are massive numbers. I think you should find a better way to cache this.

Reply:
It was discussed that we should do whatever faster is here. (https://github.com/sympy/sympy/pull/16145#issuecomment-470649409).

Reply:
We could recompute them each time as well. It depends on whether we consider this to be too much memory usage.

Reply:
If we are in favour of saving memory then we can use something like `python generators` and can generate the term whenever we required.(Will save lot of memory)
But all this process will take too much time and IMO we should give priority to time as no. of terms are limited here.

Reply:
An even number is perfect if it can be expressed as two factors: one a power of two and the other being one less than a related power of two. Except for the first 23 Mersenne prime powers it would be much faster to check the exponent:
```python
def _isperfect(n):
    if n%2 == 0:
        i = multiplicity(2, n)
        o = n >> i
        return multiplicity(2, o + 1) == i + 1
```
So I would suggest precalculating the first 23 perfect numbers and then for a given number do
```python
PERFECT = [2**(i-1)*(2**i-1) for i in MERSENNE_PRIME_EXPONENTS[:23]]
def _isperfect(n):
    global perfect
    if n % 2 == 0:
        while n > PERFECT[-1] and len(PERFECT) < len(MERSENNE_PRIME_EXPONENTS):
            i = 2**(MERSENNE_PRIME_EXPONENTS[len(PERFECT)] - 1)
            PERFECT.append(i*(2*i - 1))
        if n <= PERFECT[-1]:
            return n in PERFECT
    else:
        i,_ = integer_log(n, 2)
        if i < exponent below which all odds are known to be not perfect:
            return False
    # handle new analysis for odd or even
```

Reply:
It takes 5 seconds to initialize `PERFECT = [2**(i - 1)*(2**i - 1) for i in MERSENNE_PRIME_EXPONENTS]` on a powerful computer.I don't know what you mean by "priority to time" but that is not fast.

Reply:
I think that's the point. If we cache the values that 5 seconds only needs to be run once. But it requires storing the numbers in memory. 

Personally I think that

- the 100 or so MB that this takes isn't so bad on a modern computer. But neither is 5 seconds.
- very few people will actually call these functions. I don't think they would be used in other places in SymPy (or am I wrong?). So it doesn't actually matter that much.

Reply:
With the incremental caching it's much better.